### PR TITLE
feat: Roles & Permissions

### PR DESCRIPTION
**Pull request by kristianbinau**
## Description

- Added Roles
- Added Permissions
- Added user/roles endpoint
- Added Pivot tables
- Added Authentication w/ permissions to all endpoints
- Added Seeders

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: prisma/seed.ts [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+    let floors: Floor[][] = [];
+    let roomTypes: RoomType[][] = [];
+
+    /**
+     * Permission upserts
+     */
+    const hotelCreatePermission = await addPermission('Hotel Create')
+    const hotelUpdatePermission = await addPermission('Hotel Update');
+    const hotelDeletePermission = await addPermission('Hotel Delete');
+    const hotelFloorsBrowsePermission = await addPermission('Hotel-Floors Browse');
```
Comment (Anders164a) [inline prisma/seed.ts:21]: All browse should be changed to GetAll
All Show should be changed to Get

File: prisma/seed.ts [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+    const roomTypeUpdatePermission = await addPermission('RoomType Update');
+    const roomTypeDeletePermission = await addPermission('RoomType Delete');
+    const roomTypeRoomsBrowsePermission = await addPermission('RoomType-Rooms Browse');
+
+    const roomBrowsePermission = await addPermission('Room Browse');
+    const roomShowPermission = await addPermission('Room Show');
+    const roomCreatePermission = await addPermission('Room Create');
+    const roomUpdatePermission = await addPermission('Room Update');
+    const roomDeletePermission = await addPermission('Room Delete');
+    const rofsgjrnPermission = await addPermission('Room Delete');
```
Comment (Anders164a) [inline prisma/seed.ts:40]: ... ?

**Review by Anders164a — CHANGES_REQUESTED** [review]

File: prisma/seed.ts [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+        ),
+    ]);
+    roomTypes.push(copenhagenRoomTypes, odenseRoomTypes, aarhusRoomTypes);
+
+    /**
+     * Rooms upserts
+     */
+    await addRooms(copenhagenHotel, copenhagenFloors, copenhagenRoomTypes, 6000);
+    await addRooms(odenseHotel, odenseFloors, odenseRoomTypes, 1500);
+    await addRooms(aarhusHotel, aarhusFloors, aarhusRoomTypes, 2500);
```
Comment (kristianbinau) [inline prisma/seed.ts:268]:
```suggestion
    await Promise.all([addRooms(copenhagenHotel, copenhagenFloors, copenhagenRoomTypes, 6000),
    addRooms(odenseHotel, odenseFloors, odenseRoomTypes, 1500),
    addRooms(aarhusHotel, aarhusFloors, aarhusRoomTypes, 2500)]);
```

File: src/modules/floor/__test__/deleteFloor.test.ts [hunk 4]
```diff
@@ -63,4 +74,43 @@ describe("DELETE /api/floor/:id", () => {
             statusCode: 400,
         });
     });
+
+        it("should return status 401 when no user is provided", async () => {
```
Comment (kristianbinau) [inline src/modules/floor/__test__/deleteFloor.test.ts:78]: Formatting

File: src/plugins/jwt.ts [hunk 5]
```diff
@@ -13,16 +13,16 @@ declare module "fastify" {
     }
     interface FastifyInstance {
         authenticate(
-            request: FastifyRequest,
+            request: FastifyRequest<{ Params: any, Body: any }>,
```
Comment (kristianbinau) [inline src/plugins/jwt.ts:16]: 🥇

**Review by frederikpyt — APPROVED** [review]

**Pull request merged**